dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.SemanticKernel.Process.Internal;
+using Microsoft.SemanticKernel.Process.Runtime;
+
+namespace Microsoft.SemanticKernel;
+
+internal sealed class LocalMap : LocalStep
+{
+    private readonly HashSet<string> _mapEvents;
+    private readonly KernelProcessMap _map;
+    private readonly ILogger _logger;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="LocalMap"/> class.
+    /// </summary>
+    /// <param name="map">The <see cref="KernelProcessMap"/> instance.</param>
+    /// <param name="kernel">An instance of <see cref="Kernel"/></param>
+    internal LocalMap(KernelProcessMap map, Kernel kernel)
+        : base(map, kernel)
+    {
+        this._map = map;
+        this._logger = this._kernel.LoggerFactory?.CreateLogger(this._map.State.Name) ?? new NullLogger<LocalStep>();
+        this._mapEvents = [.. map.Edges.Keys.Select(key => key.Split('.').Last())];
+    }
+
+    /// <inheritdoc/>
+    internal override async Task HandleMessageAsync(ProcessMessage message)
+    {
+        IEnumerable values = message.GetMapInput(this._logger);
+
+        int index = 0;
+        List<(Task Task, LocalKernelProcessContext ProcessContext, MapOperationContext Context)> mapOperations = [];
+        ConcurrentDictionary<string, Type> capturedEvents = [];
+
+        try
+        {
+            foreach (var value in values)
+            {
+                ++index;
+
+                KernelProcess process = this._map.Operation.CloneProcess(this._logger);
+                MapOperationContext context = new(this._mapEvents, capturedEvents);
+#pragma warning disable CA2000 // Dispose objects before losing scope
+                LocalKernelProcessContext processContext = new(process, this._kernel, context.Filter);
+                Task processTask =
+                    processContext.StartWithEventAsync(
+                        new KernelProcessEvent
```

Comment:
so all events emitted from a map are internal right? maybe if one was marked as public at this level a warning can be raised so people know

Reply:
Correct, events specific to the map-operation (map-step edges) are internal as they target the process in which the map step is defined.  Any other proxied event will retain its original visibility.